File: cpp/src/53.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <cassert>

using namespace std;


class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_ending_here = 0, max_so_far = INT_MIN;
        
        for (int num: nums) {
            max_ending_here = max(num, max_ending_here + num);
            max_so_far = max(max_so_far, max_ending_here);
        }
        
        return max_so_far;
    }
};


int main() {
    vector<int> in = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    Solution solution = Solution();
    int ret = solution.maxSubArray(in);
    assert(ret == 6);
    cout << "Success" << endl;

    return 0;
}

File: go/src/53.go
func maxSubArray(nums []int) int {
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	ans := nums[0]

	for i := 1; i < len(nums); i++ {
		dp[i] = max(nums[i], dp[i-1]+nums[i])
		ans = max(ans, dp[i])
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}